File: src/main/java/com/chaco/algorithms/tree/minTreeDepth.java
package com.chaco.algorithms.tree;

import com.chaco.algorithms.TreeNode;

/**
 * <a href="https://leetcode.cn/problems/minimum-depth-of-binary-tree/">...</a>
 * 给定一个二叉树，找出其最小深度。
 * <p>
 * 最小深度是从根节点到最近叶子节点的最短路径上的节点数量。
 */
public class minTreeDepth {
    public static void main(String[] args) {
        System.out.println("minTreeDepth.main");
    }

    public int minDepth(TreeNode root) {
        if (null == root) {
            return 0;
        }
        if (null == root.left && null == root.right) {
            return 1;
        }
        if (null == root.left) {
            return minDepth(root.right) + 1;
        }
        if (null == root.right) {
            return minDepth(root.left) + 1;
        }
        return Math.min(minDepth(root.right), minDepth(root.left)) + 1;
    }
}

File: src/main/java/com/chaco/algorithms/tree/minTreeDepth.go
package tree

import "math"

func minDepth(root *TreeNode) int {
	if nil == root {
		return 0
	}
	if nil == root.Left && nil == root.Right {
		return 1
	}
	if nil == root.Left {
		return minDepth(root.Right) + 1
	}
	if nil == root.Right {
		return minDepth(root.Left) + 1
	}
	return int(math.Min(float64(minDepth(root.Left)), float64(minDepth(root.Right))) + 1)
}
